pkg/asset/installconfig/vsphere/validation.go

```diff
@@ -66,3 +82,99 @@ func folderExists(ic *types.InstallConfig, fldPath *field.Path) field.ErrorList
 	}
 	return nil
 }
+
+func validatevCenterVersion(vim25 *vim25.Client) error {
+	constraints, err := version.NewConstraint("<= 6.7")
+
+	if err != nil {
+		return err
+	}
+
+	vCenterVersion, err := version.NewVersion(vim25.ServiceContent.About.Version)
+	if err != nil {
+		return err
+	}
+	build, err := strconv.Atoi(vim25.ServiceContent.About.Build)
+	if err != nil {
+		return err
+	}
+
+	if constraints.Check(vCenterVersion) {
+		logrus.Warnf("The out-of-tree storage driver requires virtual hardware 15 and vSphere 6.7 U3. Current vCenter version: %s, build: %s",
+			vim25.ServiceContent.About.Version, vim25.ServiceContent.About.Build)
+
+	} else {
+		// This is the vCenter 6.7 U3 build number
+		// Anything less than this version is unsupported with the
+		// out-of-tree CSI.
+		// https://kb.vmware.com/s/article/2143838
+		// https://vsphere-csi-driver.sigs.k8s.io/supported_features_matrix.html
+		if build < 14368073 {
+			logrus.Warnf("The out-of-tree storage driver requires virtual hardware 15 and vSphere 6.7 U3. Current vCenter version: %s, build: %s",
+				vim25.ServiceContent.About.Version, vim25.ServiceContent.About.Build)
+		}
+	}
+	return nil
+}
+
+func validateESXiVersion(vim25 *vim25.Client, ic *types.InstallConfig) error {
+	cfg := ic.VSphere
+	clusterPath := fmt.Sprintf("/%s/host/%s", cfg.Datacenter, cfg.Cluster)
+	finder := find.NewFinder(vim25)
+
+	ctx, cancel := context.WithTimeout(context.TODO(), 60*time.Second)
+	defer cancel()
+
+	clusters, err := finder.ClusterComputeResourceList(ctx, clusterPath)
+	if err != nil {
+		return errors.Wrapf(err, "unable to find cluster on path: %s", clusterPath)
+	}
+
+	constraints, err := version.NewConstraint(">= 6.7")
+	if err != nil {
+		return err
+	}
+	minVerConstraints, err := version.NewConstraint("= 6.7")
+	if err != nil {
+		return err
+	}
+	hosts, err := clusters[0].Hosts(context.TODO())
+	if err != nil {
+		return errors.Wrapf(err, "unable to find hosts from cluster on path: %s", clusterPath)
+	}
+
+	for _, h := range hosts {
+		var mh mo.HostSystem
+		h.Properties(context.TODO(), h.Reference(), []string{"config.product"}, &mh)
+
+		v1, err := version.NewVersion(mh.Config.Product.Version)
+		if err != nil {
+			return err
+		}
+
+		// Check if ESXi host is greater than or equal to 6.7
+		if constraints.Check(v1) {
+			// Check if ESXi host is exactly 6.7
+			if minVerConstraints.Check(v1) {
+				build, err := strconv.Atoi(mh.Config.Product.Build)
+				if err != nil {
+					return err
+				}
+				// This is the ESXi 6.7 U3 build number
+				// Anything less than this version is unsupported with the
+				// out-of-tree CSI.
+				// https://kb.vmware.com/s/article/2143838
+				// https://vsphere-csi-driver.sigs.k8s.io/supported_features_matrix.html
+				if build < 14320388 {
```

Comment:
I was thinking some more about it and can we use apiVersion rather than build number? It looks like 6.7u3 sets API version to be `6.7.3` and I expect all earlier versions to have lower numbers than that. 

![Screenshot from 2021-09-01 13-02-18](https://user-images.githubusercontent.com/278/131723847-ec9656af-ff5c-4e88-898f-043640711037.png)

Reply:
I would rather use what I have I the PR vs the apiversion in which I have no idea what VMware could put there.

Reply:
That would be my concern as well @jcpowermac .  It appears according to the schema that it __should__ be in format x.y.z https://code.vmware.com/apis/968/vsphere .  I think the build number provides a similar level of accuracy though IMO.

Reply:
I was just talking with @jcpowermac offline and we will need similar logic in vsphere-problem-detector. I just opened - https://github.com/openshift/vsphere-problem-detector/pull/48 for reporting apiVersion string.

Reply:
It should be noted that 6.5 on the other hand is setting different apiVersion. It does not seem proper but for our purposes it works. 
![Screenshot from 2021-09-01 15-13-49](https://user-images.githubusercontent.com/278/131731800-7eb2e333-9031-4768-80b0-3ee6a42082fc.png)

Reply:
@gnufied discussed back and forth on slack in regards to this. If VMware actually followed semver for apiVersion it would have been way easier to use that field for this check. Unfortunately they are unreliable (6.5 is "6.5", 7U2 is "7.0.2.0"). Just moving forward with version and build.